bus/functools.go

```diff
@@ -0,0 +1,197 @@
+package bus
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+// Endpoints couples a consumer and a publisher to a single entity.
+type Endpoints struct {
+	consumer  *Consumer
+	publisher Publisher
+}
+
+// NewEndpoints creates the Endpoints for the given publisher and consumer. If one of the values
+// is nil, the function created by this endpoint can only be used for invoking or compute but not both.
+// If both values are set, the function will be invoked by the same process which implements it otherwise
+// you have some sort of client and server for the function.
+func NewEndpoints(consumer *Consumer, publisher Publisher) *Endpoints {
+	return &Endpoints{
+		consumer:  consumer,
+		publisher: publisher,
+	}
+}
+
+// DirectEndpoints returns endpoints which call the target function directly. You do not need
+// a running nsq for this. This can be used with unit tests. It should not be used in production
+// code because the invocation of functions will not be persistent and will be delegated to the
+// current running process by forking a gorouting to call the receiving function.
+func DirectEndpoints() *Endpoints {
+	return &Endpoints{}
+}
+
+// A Function encapsulates a Func which can be called with an argument. The invocation will be delegated through
+// nsq so multiple instances of the same function can run in different processes. Only one of them
+// will be invoked.
+type Function struct {
+	endpoints *Endpoints
+	fn        reflect.Value
+	name      string
+}
+
+type Func func(interface{}) error
+
+// Function creates a Function from the the given endpoints. The name of the function will be a
+// distributed selector for the given go function. So every function which is registered with the same
+// name can receive the invocation inside the cluster.
+// The function must be a normal go function with one parameter and one result of type error:
+//   ep := NewEndpoints(...)
+//   f, err := ep.Function("hello", func (s string) error {
+//      fmt.Printf("Hello %s\n", s)
+//      return nil
+//   })
+//   f("world"); // prints "Hello world"
+// The target function can receive structs or pointer to structs. Please notice that when using
+// `DirectEndpoints` the parameters are not marshalled/unmarshalled via JSON, so using addresses
+// can have side effects.
+func (e *Endpoints) Function(name string, fn interface{}) (Func, error) {
+	return e.function(name, "function", fn)
+}
+
+// Client returns a new function client for the function with the registered name.
+func (e *Endpoints) Client(name string) (Func, error) {
+	return e.function(name, "function", nil)
+}
+
+// Unique usese an unique, ephemeral topic so the topic will be deregisted when there is no
+// consumer any more for this function. Use this function to create a unique receiver, so function
+// invocations will not be distributed and the topic only exists as long as the registration
```

Comment:
As far as I understand, unique functions mainly (but are of course not limited to) serve as one-time result-callback. Would it be feasible to provide (optional) auto-deregistration of Unique-Functions when they have received their call?
If we use Function/Unique in potentially long running processes like web/microservices, resources will quickly accumulate if not freed regularly.

Reply:
no unique functions exists as long the process runs. so they are not only for one-time results they are a result channel which ends exact in "this" client. as long as the client calls serverside functions and wants results there should exist a result-function. one COULD create a new function for every invocation, but it would be better to have ONE result function and have some sort of unique parameter which is transported back

Reply:
Ok, this makes sense. The correlation of the requests and responses has to be done by the developer using a "correlation id" in the request/response and the service-function must be aware of that.
Maybe an examples-package might be helpful to grasp the envisioned uses and recommended patterns.